File: hello-kotlin/src/main/kotlin/server/StreamManager.kt
package server

import java.io.File
import java.time.Duration
import java.time.Instant

/** Stream lifecycle status. */
enum class StreamStatus { UPLOADING, READY, ERROR }

/** Manages all active streams and their caches. */
class StreamManager(private val cacheDir: String = "audio/output") {

    companion object {
        const val MAX_STREAMS = 1000
        private val MAX_IDLE = Duration.ofHours(24)
        private val MAX_UPLOADING = Duration.ofHours(1)
    }

    private val streams = mutableMapOf<String, Stream>()
    private val lock = Any()

    init { File(cacheDir).mkdirs() }

    fun create(streamId: String) {
        synchronized(lock) {
            if (streams.size >= MAX_STREAMS) throw IllegalStateException("Max streams ($MAX_STREAMS) reached")
            if (streams.containsKey(streamId)) throw IllegalStateException("Stream already exists: $streamId")
            val cache = MmapCache("$cacheDir/$streamId.cache")
            cache.create()
            streams[streamId] = Stream(streamId, cache)
        }
    }

    fun write(streamId: String, data: ByteArray) {
        val s = getStream(streamId)
        synchronized(s.lock) {
            s.cache.write(s.offset, data)
            s.offset += data.size
            s.lastAccess = Instant.now()
        }
    }

    fun complete(streamId: String) {
        val s = getStream(streamId)
        synchronized(s.lock) {
            s.cache.finalize(s.offset)
            s.status = StreamStatus.READY
            s.lastAccess = Instant.now()
        }
    }

    fun read(streamId: String, offset: Long, length: Int): ByteArray {
        val s = getStream(streamId)
        synchronized(s.lock) {
            s.lastAccess = Instant.now()
            return s.cache.read(offset, length)
        }
    }

    fun delete(streamId: String) {
        synchronized(lock) {
            streams.remove(streamId)?.cache?.close()
        }
    }

    fun markError(streamId: String) {
        synchronized(lock) {
            streams[streamId]?.status = StreamStatus.ERROR
        }
    }

    fun statusOf(streamId: String): Pair<StreamStatus, Long> {
        val s = getStream(streamId)
        synchronized(s.lock) { return s.status to s.offset }
    }

    fun list(): List<String> = synchronized(lock) { streams.keys.toList() }

    fun cleanup(): Int {
        synchronized(lock) {
            val now = Instant.now()
            val toRemove = streams.filter { (_, s) ->
                val idle = Duration.between(s.lastAccess, now)
                idle > MAX_IDLE
                    || (s.status == StreamStatus.UPLOADING && idle > MAX_UPLOADING)
                    || s.status == StreamStatus.ERROR
            }.keys.toList()
            toRemove.forEach { streams.remove(it)?.cache?.close() }
            return toRemove.size
        }
    }

    data class Stats(val total: Int, val uploading: Int, val ready: Int, val error: Int)

    fun stats(): Stats {
        synchronized(lock) {
            var u = 0; var r = 0; var e = 0
            for (s in streams.values) {
                when (s.status) {
                    StreamStatus.UPLOADING -> u++
                    StreamStatus.READY -> r++
                    StreamStatus.ERROR -> e++
                }
            }
            return Stats(streams.size, u, r, e)
        }
    }

    // ── Internal ────────────────────────────────────────────────────────

    private fun getStream(streamId: String): Stream =
        synchronized(lock) { streams[streamId] ?: throw NoSuchElementException("Stream not found: $streamId") }

    class Stream(
        val id: String,
        val cache: MmapCache,
        var offset: Long = 0,
        val created: Instant = Instant.now(),
        var lastAccess: Instant = Instant.now(),
        var status: StreamStatus = StreamStatus.UPLOADING,
        val lock: Any = Any(),
    )
}

File: hello-csharp/src/Server/StreamManager.cs
namespace AudioStreamServer;

/// <summary>Stream lifecycle status.</summary>
public enum StreamStatus { Uploading, Ready, Error }

/// <summary>Manages all active streams and their caches.</summary>
public sealed class StreamManager
{
    public const int MaxStreams = 1000;
    private static readonly TimeSpan MaxIdleTime = TimeSpan.FromHours(24);
    private static readonly TimeSpan MaxUploadingTime = TimeSpan.FromHours(1);

    private readonly string _cacheDir;
    private readonly Dictionary<string, Stream> _streams = new();
    private readonly object _lock = new();

    public StreamManager(string cacheDir = "audio/output")
    {
        _cacheDir = cacheDir;
        Directory.CreateDirectory(cacheDir);
    }

    public void Create(string streamId)
    {
        lock (_lock)
        {
            if (_streams.Count >= MaxStreams)
                throw new InvalidOperationException($"Max streams ({MaxStreams}) reached");
            if (_streams.ContainsKey(streamId))
                throw new InvalidOperationException($"Stream already exists: {streamId}");

            var cache = new MmapCache(Path.Combine(_cacheDir, $"{streamId}.cache"));
            cache.Create();
            _streams[streamId] = new Stream(streamId, cache);
        }
    }

    public void Write(string streamId, byte[] data)
    {
        var stream = GetStream(streamId);
        lock (stream.Lock)
        {
            stream.Cache.Write(stream.Offset, data);
            stream.Offset += data.Length;
            stream.LastAccess = DateTime.UtcNow;
        }
    }

    public void Complete(string streamId)
    {
        var stream = GetStream(streamId);
        lock (stream.Lock)
        {
            stream.Cache.Finalize(stream.Offset);
            stream.Status = StreamStatus.Ready;
            stream.LastAccess = DateTime.UtcNow;
        }
    }

    public byte[] Read(string streamId, long offset, int length)
    {
        var stream = GetStream(streamId);
        lock (stream.Lock)
        {
            stream.LastAccess = DateTime.UtcNow;
            return stream.Cache.Read(offset, length);
        }
    }

    public void Delete(string streamId)
    {
        lock (_lock)
        {
            if (_streams.Remove(streamId, out var stream))
            {
                stream.Cache.Dispose();
            }
        }
    }

    public void MarkError(string streamId)
    {
        lock (_lock)
        {
            if (_streams.TryGetValue(streamId, out var stream))
                stream.Status = StreamStatus.Error;
        }
    }

    public (StreamStatus status, long size) StatusOf(string streamId)
    {
        var stream = GetStream(streamId);
        lock (stream.Lock)
        {
            return (stream.Status, stream.Offset);
        }
    }

    public string[] List()
    {
        lock (_lock)
        {
            return _streams.Keys.ToArray();
        }
    }

    public int Cleanup()
    {
        lock (_lock)
        {
            var now = DateTime.UtcNow;
            var toRemove = _streams.Where(kv =>
            {
                var s = kv.Value;
                var idle = now - s.LastAccess;
                return idle > MaxIdleTime
                    || (s.Status == StreamStatus.Uploading && idle > MaxUploadingTime)
                    || s.Status == StreamStatus.Error;
            }).Select(kv => kv.Key).ToList();

            foreach (var id in toRemove)
            {
                if (_streams.Remove(id, out var stream))
                    stream.Cache.Dispose();
            }
            return toRemove.Count;
        }
    }

    public (int total, int uploading, int ready, int error) Stats()
    {
        lock (_lock)
        {
            int u = 0, r = 0, e = 0;
            foreach (var s in _streams.Values)
            {
                switch (s.Status)
                {
                    case StreamStatus.Uploading: u++; break;
                    case StreamStatus.Ready: r++; break;
                    case StreamStatus.Error: e++; break;
                }
            }
            return (_streams.Count, u, r, e);
        }
    }

    // ── Internal ────────────────────────────────────────────────────────

    private Stream GetStream(string streamId)
    {
        lock (_lock)
        {
            return _streams.TryGetValue(streamId, out var s)
                ? s
                : throw new KeyNotFoundException($"Stream not found: {streamId}");
        }
    }

    /// <summary>Internal stream state.</summary>
    internal sealed class Stream
    {
        public string Id { get; }
        public MmapCache Cache { get; }
        public long Offset { get; set; }
        public DateTime Created { get; }
        public DateTime LastAccess { get; set; }
        public StreamStatus Status { get; set; }
        public object Lock { get; } = new();

        public Stream(string id, MmapCache cache)
        {
            Id = id;
            Cache = cache;
            Created = DateTime.UtcNow;
            LastAccess = DateTime.UtcNow;
            Status = StreamStatus.Uploading;
        }
    }
}
